Add email services DNS records

gius80 opened the pull request:
Closes [DOTDASH-265](https://linear.app/a8c/issue/DOTDASH-265/add-txt-email-verification-module-in-dns-list-page)

## Proposed Changes

This PR implements the email service providers module in the DNS list page, allowing users to automatically generate DNS records for major providers including Google Workspace, Office 365, Apple Mail, and Zoho Mail.

The form requires the verification token provided by the selected service. Once submitted, the DNS template is automatically applied on the backend, generating all the necessary DNS records.
![Screenshot 2025-08-14 alle 15 30 04](https://github.com/user-attachments/assets/046a5f72-04db-46c7-b35c-1c18a2bead4e)

> [!NOTE]  
> Since there's no an Accordion component yet, the card is always expanded.

## Why are these changes being made?
This component is part of the HD Domains projects.

## Testing Instructions
- Test the form using real or fake verification tokens.
- Verify that token validation works as expected.
- Ensure that the submit button remains disabled if the token is invalid.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

gius80 committed:
Add email services verification records

gius80 committed:
Add dns template mutation and constant

gius80 committed:
Add notices

gius80 committed:
Minor fixes

gius80 committed:
Fine tuning

gius80 committed:
Fine tuning

gius80 committed:
Disable button when the form is not valid

bogiii commented on client/dashboard/app/queries/domain-dns-records.ts:
```diff
@@ -42,3 +43,21 @@ export const domainDnsEmailMutation = ( domainName: string ) =>
+			variables: {
+				domain: string;
+				token: string;
+				mxdata?: string;
```
I can see this block in multiple places. Can we create a `Variable` type for:
```
    domain: string;
    token: string;
    mxdata?: string;
```

bogiii commented on client/dashboard/domains/domain-dns/email-setup-form.tsx:
```diff
@@ -0,0 +1,147 @@
+	modifyVariables?: ( variables: { token: string; domain: string; mxdata?: string } ) => {
+		token: string;
+		domain: string;
+		mxdata?: string;
```
Variable type

bogiii commented on client/dashboard/domains/domain-dns/email-setup-form.tsx:
```diff
@@ -0,0 +1,147 @@
+		let variables: {
+			token: string;
+			domain: string;
+			mxdata?: string;
```
Variable type

bogiii commented on client/dashboard/domains/domain-dns/email-setup-form.tsx:
```diff
@@ -0,0 +1,147 @@
+					} );
+				},
+				onSettled: () => {
+					setIsPending( false );
```
The manual handling of `isPending` state is redundant and not the best practice here because `useMutation` from React Query already provides this state for us through `mutation.isPending`

fushar replied:
+1

fushar commented on client/dashboard/app/queries/domain-dns-records.ts:
```diff
@@ -42,3 +43,21 @@ export const domainDnsEmailMutation = ( domainName: string ) =>
+
+export const domainDnsApplyTemplateMutation = ( domainName: string ) =>
+	mutationOptions( {
+		mutationFn: ( params: {
```
[nit] I think we don't need to explicitly use named parameter `param` here if it's never being used (we can just spread everything).

fushar commented on client/dashboard/data/domain-dns-records.ts:
```diff
@@ -60,3 +60,21 @@ export function restoreDefaultEmailRecords( domainName: string ): Promise< void
+	domainName: string,
+	provider: string,
+	service: string,
+	variables: {
```
We should extract this to a type, maybe `DnsTemplateVariables`?

fushar commented on client/dashboard/domains/domain-dns/email-setup.tsx:
```diff
@@ -0,0 +1,110 @@
+			name: 'google-workspace',
+			title: __( 'Google Workspace' ),
+			content: (
+				<div className="email-setup__tab-content">
```
We don't seem to have CSS for this? Then we can remove it for now 😄

fushar commented on client/dashboard/domains/domain-dns/email-setup-form.tsx:
```diff
@@ -0,0 +1,147 @@
+interface EmailSetupFormProps {
+	description: string;
+	label: string;
+	modifyVariables?: ( variables: { token: string; domain: string; mxdata?: string } ) => {
```
[super nit] a more common term for this use case is `transform` instead of `modify` 😄

bogiii commented on client/dashboard/domains/domain-dns/email-setup-form.tsx:
```diff
@@ -0,0 +1,147 @@
+		},
+	];
+
+	const canSubmit = isItemValid( formData, fields, form );
```
I did not know about this method. I'm going to implement on my Namer Servers form. Thanks for that! 😄

bogiii replied:
FYI: @hambai ☝️

bogiii commented on client/dashboard/domains/domain-dns/email-setup.tsx:
```diff
@@ -0,0 +1,110 @@
+			name: 'google-workspace',
+			title: __( 'Google Workspace' ),
+			content: (
+				<div className="email-setup__tab-content">
```
Do we need this div wrapper?

gius80 replied:
Nope 😅

bogiii commented on client/dashboard/domains/domain-dns/email-setup.tsx:
```diff
@@ -0,0 +1,110 @@
+						className="email-setup__tabs"
+						orientation={ isMobile ? 'horizontal' : 'vertical' }
+					>
+						{ ( tab ) => <div className="email-setup__tab-panel">{ tab.content }</div> }
```
Do we need this div wrapper at all?

bogiii reviewed (changes requested):
Form submission doesn't trigger any process. See video:

https://github.com/user-attachments/assets/87a9f007-6491-4968-a0b9-1d91cf3a929e

Just thinking, should we clear the input while switching between tabs?

https://github.com/user-attachments/assets/fa3dee07-dbb6-4e9e-854d-29f92027bb24

gius80 committed:
Address code review suggestions

gius80 committed:
Minor fix

gius80 committed:
Add key to panels to reset form data

gius80 commented:
> Form submission doesn't trigger any process. See video:
I'm stupid, I don't know why but I removed `type=sumbit` 😅

> Just thinking, should we clear the input while switching between tabs?
I agree! I’ve added keys to the tabs so the component re-renders and the form gets reset.

bogiii reviewed (changes requested):
After submission is successful, we should reset the form to the initial state without validation errors. I had impression that something is wrong/failed.


https://github.com/user-attachments/assets/05f07a87-e7ff-48d3-80e6-254f4b8aae93

p-jackson commented on client/dashboard/domains/domain-dns/email-setup.tsx:
```diff
@@ -0,0 +1,106 @@
+				<VStack spacing={ 5 }>
+					<SectionHeader
+						title={ __( 'Email setup' ) }
+						description={ __( 'Set up an existing email service for this domain' ) }
```
```suggestion
						description={ __( 'Set up an existing email service for this domain.' ) }
```

p-jackson commented on client/dashboard/domains/domain-dns/email-setup.tsx:
```diff
@@ -0,0 +1,106 @@
+					) }
+					label="Zoho Mail"
+					pattern={ /^zb\w{1,100}$/ }
+					placeholder="zb..."
```
```suggestion
					placeholder="zb…"
```

p-jackson commented on client/dashboard/domains/domain-dns/email-setup.tsx:
```diff
@@ -0,0 +1,106 @@
+						} )
+					}
+					pattern={ /^MS=ms\d{8}$/ }
+					placeholder="MS=ms..."
```
```suggestion
					placeholder="MS=ms…"
```

p-jackson commented on client/dashboard/domains/domain-dns/email-setup.tsx:
```diff
@@ -0,0 +1,106 @@
+					description={ __(
+						'Paste the verification token provided by Office 365 for the TXT record.'
+					) }
+					label="Office 365"
```
Are these labels translatable in the old interface? It seems odd that `title` would be translated but `label` would not. I know they're brand names, but presumably the translators are handling that case correctly for the title 🤔

gius80 replied:
My bad, the title shouldn't be translatable as well 😅

p-jackson commented on client/dashboard/domains/domain-dns/email-setup-form.tsx:
```diff
@@ -0,0 +1,139 @@
+			placeholder,
+			isValid: {
+				custom: ( item ) => {
+					return pattern.test( item.record ) ? null : __( 'Invalid verification record format' );
```
```suggestion
					return pattern.test( item.record ) ? null : __( 'Invalid verification record format.' );
```

p-jackson commented on client/dashboard/domains/domain-dns/email-setup-form.tsx:
```diff
@@ -0,0 +1,139 @@
+	const canSubmit = isItemValid( formData, fields, form );
+
+	return (
+		<div className="email-setup-form" style={ { marginTop: '28px' } }>
```
It's too bad about this custom margin. The core `<TabPanel>` doesn't provide any spacing of its own, it is all up to the tab content to do its own spacing. Which I guess makes sense, the content probably has a better idea of what is appropriate in each situation.

In this example from storybook none of this spacing is coming from `<TabPanel>`, it is all just from the paragraph margins.

![CleanShot 2025-08-18 at 21.15.29@2x.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/roXrVU9XgOXyAiRGxI4s/37f30147-39ae-4388-a405-02b2ac01575e.png)

Not a blocker for this PR. @lucasmendes-design this feels like something to consider because I can imagine inconsistencies growing as tab interfaces get added.

p-jackson commented on client/dashboard/components/section-header/style.scss:
```diff
@@ -36,6 +36,10 @@
+
+	.dashboard-section-header.is-level-4 & {
+		@include heading-medium();
+	}
```
Unless there's a mockup we're following, I don't think we need to add new heading levels. Since we're flying blind a little bit at the moment without too much design input, I think it would be better to try and build what we can with the design elements we just happen to have at hand.

gius80 replied:
![Screenshot 2025-08-18 alle 13 33 52](https://github.com/user-attachments/assets/2931b276-a974-4ddd-bb30-a1d98eb4b767)
I'm following this one, where the title is 13px. But if @ollierozdarz agrees, we can use level-3 (15 px) instead

gius80 committed:
Address code review suggestion

gius80 committed:
Brand title should not be translatable

arthur791004 commented on client/dashboard/domains/domain-dns/email-setup-form.tsx:
```diff
@@ -0,0 +1,142 @@
+	const canSubmit = formData.record && isItemValid( formData, fields, form );
+
+	return (
+		<div className="email-setup-form" style={ { marginTop: '28px' } }>
```
Do we need the className 🤔

gius80 replied:
Nope 😅

arthur791004 commented on client/dashboard/domains/domain-dns/email-setup.tsx:
```diff
@@ -0,0 +1,106 @@
+	];
+
+	return (
+		<Card className="email-setup">
```
Do we need the className 🤔

gius80 committed:
Remove classnames

gius80 committed:
Address PR review feedback

gius80 committed:
Fix route import
